internal/ordered/unmarshal.go

```diff
@@ -0,0 +1,375 @@
+	}
+	return nil
+}
+
+// decodeInto loads the contents of the map into the target (pointer to struct).
+// It behaves sort of like `yaml.Node.Decode`:
+//
+//   - If target is a map type with string keys, it unmarshals its contents
+//     elementwise, with values passed through Unmarshal.
+//   - If target is *struct{...}, it matches keys to exported fields either
+//     by looking at `yaml` tags, or using lowercased field names.
+//   - If a field has a yaml:",inline" tag, it copies any leftover values into
+//     that field, which must have type map[string]any or any. (Structs are not
+//     supported for inline.)
+func (m *Map[K, V]) decodeInto(target any) error {
+	tm, ok := any(m).(*Map[string, any])
+	if !ok {
+		return fmt.Errorf("%w: cannot unmarshal from %T, want K=string, V=any", ErrIncompatibleTypes, m)
+	}
+
+	// Work out the kind of target being used.
+	// Dereference the target to find the inner value, if needed.
+	targetValue := reflect.ValueOf(target)
+	var innerValue reflect.Value
+	switch targetValue.Kind() {
+	case reflect.Pointer:
+		// Passed a pointer to something.
+		if targetValue.IsNil() {
+			return ErrIntoNil
+		}
+		innerValue = targetValue.Elem()
+
+	case reflect.Map:
+		// Passed a map directly.
+		innerValue = targetValue
+		if innerValue.IsNil() {
+			return ErrIntoNil
+		}
+
+	default:
+		return fmt.Errorf("%w: cannot unmarshal %T into %T, want map or *struct{...}", ErrIncompatibleTypes, m, target)
+	}
+
+	switch innerValue.Kind() {
+	case reflect.Map:
+		// Process the map directly.
+		mapType := innerValue.Type()
+		// For simplicity, require the key type to be string.
+		if keyType := mapType.Key(); keyType.Kind() != reflect.String {
+			return fmt.Errorf("%w for map key: cannot unmarshal %T into %T", ErrIncompatibleTypes, m, target)
+		}
+
+		// If target is a pointer to a nil map (with type), create a new map.
+		if innerValue.IsNil() {
+			innerValue.Set(reflect.MakeMapWithSize(mapType, tm.Len()))
+		}
+
+		valueType := mapType.Elem()
+		return tm.Range(func(k string, v any) error {
+			nv := reflect.New(valueType)
+			if err := Unmarshal(v, nv.Interface()); err != nil {
+				return err
+			}
+			innerValue.SetMapIndex(reflect.ValueOf(k), nv.Elem())
+			return nil
+		})
+
+	case reflect.Struct:
+		// The rest of the method is concerned with this.
+	default:
+		return fmt.Errorf("%w: cannot unmarshal %T into %T", ErrIncompatibleTypes, m, target)
+	}
+
+	// These are the (accessible by reflection) fields it has.
+	// This includes non-exported fields.
+	fields := reflect.VisibleFields(innerValue.Type())
+
+	var inlineField reflect.StructField
+	outlineKeys := make(map[string]struct{})
+
+	for _, field := range fields {
+		// Skip non-exported fields. This is conventional *and* correct.
+		if !field.IsExported() {
+			continue
+		}
+
+		// No worries if the tag is not there - apply defaults.
+		tag, _ := field.Tag.Lookup("yaml")
```

Comment:
were we to export this into its own library, would we want to give this its own struct tag? or is it too tied into the yaml.v3 ecosystem?

Reply:
Yes, I think having its own tag would be an appropriate next step.